### ISSUE TITLE
[Suggestion and a bit of code] Forcing symmetry (mgrid adjustments)

### ISSUE THREAD

**Issue by torridgristle** [lucidrains/deep-daze#12]
In the original (CLIP & gradient ascent) notebook, there was a function named get_mgrid what set up a 2D tensor from -1 to 1 with the desired image size as the steps. I was able to modify this to enforce various layouts, such as horizontal symmetry. Admittedly, not super clear on how to do what I want with it since used to thinking about things like this as vectors in GLSL where you can just texcoord.x = abs((texcoord.x * 2.0) - 1.0); and call it a day.

For reference, my adjustment to make the horizontally mirrored canvas was to add this after `mgrid = mgrid.reshape(-1,dim)`:
`mgrid2 = torch.sub(torch.abs(torch.mul(mgrid, 2.0)), 1)` to change -1, 0, 1 to -2, 0, 2, then to 2, 0, 2, then to 1, -1, 1.
`mgrid = torch.add(torch.mul(mgrid, torch.tensor([1.0, 0.0])), torch.mul(mgrid2, torch.tensor([0.0, 1.0])))` in an attempt to mask out the X and Y coordinates and insert the mirrored one into the x coordinate, though looking at this it places it in the second slot so I have no idea why that even worked if X comes before Y. I have very little knowledge on how these things are stored / accessed / laid out.

An option for different layouts like this would be useful if not simply entertaining for aesthetic purposes. Horizontal symmetry, vertical symmetry, radial symmetry (couldn't figure out how to make it work, kept ending up as straight lines in X or Y direction when I tried to convert X,Y to Angle, Distance), polar layout, cylindrical layout, arc layout n-star symmetry, etc.

**Comment by lucidrains** [comment]
@torridgristle interesting, let me look later this week if i can expose a way to specify custom mgrids, passed in during initialization
